### ISSUE TITLE
Update AP Patches for DSi Mode

### ISSUE THREAD

**Issue by SNBeast** [DS-Homebrew/TWiLightMenu#1422]
**Is your feature request related to a problem? Please describe.**
With the implementation of a functional DSi mode in nds-bootstrap, it has come to light that some games have DSi mode-specific anti-piracy [AP] and/or their AP patches designed for DS mode break. These games are or are close to being impossible to play in DSi mode as a result.

**Describe the solution you'd like**
Updated AP patches when available.

**Describe alternatives you've considered**
For now, the alternative is running these games in DS mode.

**Additional context**
Known affected games (subject to change upon discovery and resolution):
Pokémon Conquest - English: Whitescreens immediately.
Pokémon Black Version 2 - All releases: See Pokémon White Version 2
Pokémon White Version 2 - All releases: You can load the main menu and its submenus/modes but entering the game world either by continuing or by the end of the professor's intro for a new game whitescreens.

**Comment by Kalanyr** [comment]
The Whitescreens on Pokemon Black2/White2 are apparently caused by the AP fixes, they work without APfixes but that means the AP is active. 

Someone should check if that's the case with Fossil Fighter Champions / Pokemon Conquest (unlikely, I believe the white screen is the same AP sympton as happend on the NDS).

**Comment by Squall-Leonhart** [comment]
> The Whitescreens on Pokemon Black2/White2 are apparently caused by the AP fixes, they work without APfixes but that means the AP is active.

And this won't be a major issue if nds bootstrap is at the point where it can fool games now, does the AP actually trigger as known to?

**Comment by Kalanyr** [comment]
Yes, the 0 Exp AP at least is still reported not sure about the other
effects.

**Comment by pistonfish** [comment]
> Yes, the 0 Exp AP at least is still reported not sure about the other effects.

The AP added some crashes which are present when loading the game in DSi-mode without apfix.

> Pokémon Black Version 2 - All releases: See Pokémon White Version 2
> Pokémon White Version 2 - All releases: You can load the main menu and its submenus/modes but entering the game world either by continuing or by the end of the professor's intro for a new game whitescreens.

The German version actually crashes with a black screen when loading the save but that seems to also happen because of the apfix.

**Comment by RocketRobz** [comment]
5bb2b74a has fixed the crash issue in the Japan/USA/Europe versions of Pokémon Black Version 2.
This was done by relocating the AP-fix code from the secure area to the unused debug font area.

@R-YaTian:
```
enler(BW2 Chinese translation's coder) found that BW2 will check AP-Patch's address
He said that it will check 0x2000800 on ds mode but 0x2004800 on dsi mode.
```

**Comment by pistonfish** [comment]
That's great, but other european versions (like the German or French version) still seem to use the old apfix.

**Comment by Squall-Leonhart** [comment]
> That's great, but other european versions (like the German or French version) still seem to use the old apfix.

there are no such versions,  
French and Deutsche black 2/white 2 was included in the EUSA release which is covered by the changes Rob has made.

**Comment by pistonfish** [comment]
> there are no such versions,
> French and Deutsche black 2/white 2 was included in the EUSA release which is covered by the changes Rob has made, if your games not working its because your file already has the no longer compatible AP fix hard patched.

My German version is a clean cartridge rip. It will start in DSi-mode when launching without TwilightMenu's apfixes but it will crash when using them.

I tried using the latest nightly and built the apfix.pck from the latest commit.

**Comment by Squall-Leonhart** [comment]
> My German version is a clean cartridge rip. It will start in DSi-mode when launching without TwilightMenu's apfixes but it will crash when using them.

Sounds wrong, the only way to get single locale nds games is via eshop.

**Comment by Lillievela** [comment]
@Squall-Leonhart if you would actually look at the compatibility list on google spreadsheets you would know that those version do in fact exist. IR*O (multiregion) is the base version for every version, IR*S is for the spanish versions, IR*D is german, IR*I is italian, and IR*F is french. Also the eshop only shows games for your specific region, meaning an italian game wont show on an NA console (also nds games arent on the eshop, only dsiware is)

**Comment by Squall-Leonhart** [comment]
The compatibility list is irrelevant.

**Comment by pistonfish** [comment]
> nintendo never produced single language cartridges for the pokemon series in these regions since they were already covered by the EUSA cartridge. they are third party fakes and stripped down from the official production rom.

They actually did. Nintendo never released multilang Pokémon games in europe until X and Y

**Comment by hansoljin** [comment]
Here is a list of games, listed from the DAT-o-MATIC database. This should clearly outline that there are, indeed, language variants of the games.

Spain
https://datomatic.no-intro.org/index.php?page=show_record&s=28&n=6598

France
https://datomatic.no-intro.org/index.php?page=show_record&s=28&n=6110

Germany
https://datomatic.no-intro.org/index.php?page=show_record&s=28&n=6111

Italy
https://datomatic.no-intro.org/index.php?page=show_record&s=28&n=6544

With that in mind, due to the ROM files being different, AP patches for each of these are not developed yet. The only AP patch present for Pokemon Black 2 is for Japanese and English games, while the other languages do not have such a patch. It will be added soon.


This issue is not meant to be Black 2 specific, and discussion about the legitimacy of a cartridge is out of scope for the issue. Please stay on topic.

**Comment by PixelateVision** [comment]
Is White 2's now updated as well? Or just Black 2 at the moment?

**Comment by RocketRobz** [comment]
Just Black 2. White 2 will be worked on after the remaining patches for Black 2 are fixed and pushed.

**Comment by KururinX** [comment]
Poket monster black 2 (JP)

CRC32:29c68b2b
MD5:ecc64d4b1d9d9d161d55ecb23db6fa64
SHA1:c55fe4412427e63ef28186af53ec114c7fdd9a0a

Freeze without starting (blackout)

- - - - - - - - - - - -
Twilight Menu Ver:3a7ca940
Bootstrap Ver:0.39.0


findModuleParamsOffset:
Module params offset restored: 02004FCC
sdk_version: 0503757C
arm9 is compressed
findLtdModuleParamsOffset:
Ltd module params offset restored: 02004F8C
arm9i is compressed

old hi heap end pointer: 048020BE

new hi heap pointer: 048020BB

Hi Heap Shrink Sucessfull

patchCardNds

[SDK 5]

swi12 location : 02004678

dsiModeCheck location : 02004EB4

cardIrqEnable location : 02079B50

startOffset : 02004000

cardRead location : 0206F794
02EF4000

cardRomInit location : 0206FB34

cardHashInit location : 02467524

cardPullOut location : 0206FBDC

cardId location : 0206F674

cardReadDma location : 0206F948


NDMA CARD READ METHOD ACTIVE
cardEndReadDma location : 0206F934

findMpuInitCacheOffset:
Mpu init cache found: 02004B34

Mpu2 init: 02004B20

reset location : 0207BF7C

ERR_NONE

SCFG_EXT location : 02380258

cardIrqEnable location : 023848C4

Relocation src: 02380644

Arm7 (patch v5)
JumpTableFunc: 0238FB14
card read:	0238FAFC
Eeprom read:	0238FB20
Eeprom page write:	0238FB38
Eeprom page prog:	0238FB50
Eeprom verify:	0238FB68
Eeprom page erase:	0238FB7C
swiHalt location : 023807A4

swi12 location : 02383BA4

swiGetPitchTable location : 02383BB8

sdCardReset location : 02EB3824

ERR_NONE

hookNdsRetailArm7
hookLocation arm7: 02391918

ERR_NONE
hookLocation arm9: 02FE0020

AP-fix found and applied
chipID: C001FEC2

- - - - - - - - - - - -
Twilight Menu Ver:3a7ca940
Bootstrap Ver:4134d9ae


findModuleParamsOffset:
Module params offset found: 02004FCC
sdk_version: 0503757C
arm9 is compressed
findLtdModuleParamsOffset:
Ltd module params offset found: 02004F8C
arm9i is compressed

findHeapPointer2Offset:
0207AFB8
old hi heap end pointer: 048020BE

new hi heap pointer: 048020BB

Hi Heap Shrink Sucessfull

patchCardNds

[SDK 5]

findSwi12Offset:
swi 0x12 call found

findDsiModeCheckOffset
DSi mode check found

swi12 location : 02004678

dsiModeCheck location : 02004EB4

findCardIrqEnableOffset:
irq enable not found
irq enable thumb alt found

cardIrqEnable location : 02079B50

startOffset : 02004000

findCardReadEndOffsetThumb5Type1:
ARM9 Card read end SDK 5 thumb (type 1) found: 0206F85C

findCardReadStartOffsetThumb5Type1:
ARM9 Card read start SDK 5 thumb alt (type 0) found

cardRead location : 0206F794
02EF4000

findCardRomInitOffsetThumb
Card ROM init found

cardRomInit location : 0206FB34

findCardHashInitOffsetThumb
Card hash init found

cardHashInit location : 02467524

findCardPullOutOffsetThumbType1:
Card pull out handler SDK 5 thumb alt (type 1) found

cardPullOut location : 0206FBDC

findCardIdEndOffsetThumb:
Card ID end thumb found: 0206F6B8

findCardIdStartOffsetThumb:
Card ID start thumb not found
Card ID start thumb alt 1 not found
Card ID start thumb alt 2 not found
Card ID start thumb alt 3 found

cardId location : 0206F674

findCardReadDmaEndOffset:
Card read DMA end found: 0206F9C0

findCardReadDmaStartOffsetThumb:
Card read DMA start thumb SDK < 3 not found
Card read DMA start thumb SDK >= 3 found

cardReadDma location : 0206F948

findCardEndReadDma
findCardEndReadDmaSdk5
cardEndReadDma found


NDMA CARD READ METHOD ACTIVE
cardEndReadDma location : 0206F934

findMpuStartOffset:
Mpu init found: 02004AD4

findMpuDataOffset:
Mpu data found: 02004B90

findMpuInitCacheOffset:
Mpu init cache found: 02004B34

Mpu2 init: 02004B20

findResetOffset
Reset found: 0207BF7C
Reset constant found: 0207C018
Reset found

reset location : 0207BF7C

ERR_NONE

a7_findScfgExtOffset:
SCFG_EXT found

SCFG_EXT location : 02380258

findSleepPatchOffset:
Sleep patch found: 0238D250

findCardIrqEnableOffset:
irq enable not found
irq enable alt found

cardIrqEnable location : 023848C4

findCardCheckPullOutOffset:
Card check pull out not found

Relocation src: 02380644

Arm7 (patch v5)
JumpTableFunc: 0238FB14
card read:	0238FAFC
Eeprom read:	0238FB20
Eeprom page write:	0238FB38
Eeprom page prog:	0238FB50
Eeprom verify:	0238FB68
Eeprom page erase:	0238FB7C
findSwiHaltOffset:
swiHalt call found

swiHalt location : 023807A4

findSwi12Offset:
swi 0x12 call found

findSwiGetPitchTableOffset:
swiGetPitchTable call SDK 5 found

swi12 location : 02383BA4

swiGetPitchTable location : 02383BB8

findSdCardResetOffset:
SD Card reset found

sdCardReset location : 02EB3824

ERR_NONE

hookNdsRetailArm7
hookLocation arm7: 02391918

ERR_NONE
handlerStartSig
0209DDA8
handlerEndSig
0209DDE8
tableAddr
02FE0020
returnAddr
01FF8CE0
hookLocation arm9: 02FE0020

AP-fix found and applied
chipID: C001FEC2

**Comment by RocketRobz** [comment]
The logs aren't relevant in that case.

**Comment by Yosaboy** [comment]
Hello. I don't know if I should open another issue or if it is related. I have a french clean ripped rom of Heartgold. I followed a tutorial to change four values in an hex editor to make the game 60fps. The apfix doesn't seem to work on the rom (randomn freezes, the Pokeballs keep turning while inside a trainer battle). I have a US rom wich as an AP Fix patch applied and it works normaly. Are modified roms unable to be AP patched by NDS Bootsrap? If so, can an AP Fix be directed to a game through name-changes? I Can't find any french AP Patch to test if it works as intended with it. Running on O3DS, latest versions of Twllight Menu ans NDS Bootsrap, tried every settings and nightly builds. I will try with a non modified french rom sometimes this week to see if it freezes still but wanted to let tout guys know of this

**Comment by RocketRobz** [comment]
Yes, the .ips file can be named the same as a ROM file, but with .ips added to the end.

**Comment by Yosaboy** [comment]
Thanks for the answer. Tried renaming the correspondant .ips file the same as my rom (tried both "name.ips" and "name.nds.ips") none worked, seems like it doesn't want to apply the patch

**Comment by RocketRobz** [comment]
If you're using the latest release, the .ips goes in `sd:/_nds/TWiLightMenu/apfix/`.
If it's the latest nightly, it goes in `sd:/_nds/TWiLightMenu/extras/apfix/`.

**Comment by Yosaboy** [comment]
Just tried what you said, didn't work. Tried with a fresch french rom with no modification, it works. Tried the tutorial on a EUR rom wich was not AP patched it worked. I even tried to directly apply the .ips patch with Lunar IPS the game didn't launch (Infinite white screen). I first thought the modified rom was the problem but if it works on an EUR rom I don't think that's the case. So I guess ether the french AP Fix doesn't work on my modified rom, or the french rom isn't structured the same as other roms. Followed this tutorial by the way : https://www.pokecommunity.com/showthread.php?t=378618
Guess I'll have to play with the EUR game, wich language can not be changed to french for some weird reason. Thanks for the help keep up the great work!

**Comment by RocketRobz** [comment]
The black screen issue in *Fossil Fighters Champions* is worked around by using a DSi-Exclusive/DSiWare game set as a donor ROM.
I recall trying to fix the existing AP-fix for DSi mode, but the game still crashed, so a donor ROM requirement will need to be added.

As this is the game on the list with a fix found, this issue is now closed, but will be pinned, in case more games are found.

**Comment by RocketRobz** [comment]
Actually, it seems the issue with *Fossil Fighters Champions* isn't AP-related. This DSi-Enhanced game is one of the few to use a THUMB-compiled ARM7 (and ARM7i) binary.
Another game to use one is *The Penguins of Madagascar: Dr. Blowhole Returns Again!*.
Both encounter the issue where the games would crash due to the save code being unusable due to the code branches being too long.

So, it seems what we thought was an AP issue for a while, was an nds-bootstrap issue.
